### PR TITLE
JDK-8287011: Improve container information

### PR DESCRIPTION
**Pull request by MBaesken**
The change adds some potentially interesting cgroup v1 and v2 metrics to hs_err / hs_info files.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287011](https://bugs.openjdk.org/browse/JDK-8287011): Improve container information

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**) ⚠️ Review applies to [77c52cdf](https://git.openjdk.org/jdk/pull/8991/files/77c52cdf6a2a4f06c625f0d88026f04ed4f47001)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8991/head:pull/8991` \
`$ git checkout pull/8991`

Update a local copy of the PR: \
`$ git checkout pull/8991` \
`$ git pull https://git.openjdk.org/jdk pull/8991/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8991`

View PR using the GUI difftool: \
`$ git pr show -t 8991`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8991.diff">https://git.openjdk.org/jdk/pull/8991.diff</a>

</details>

### CODE REVIEW

**Review by jerboaa — COMMENTED** [review]
A couple of comments:
- This introduces a couple of version specific info which is stubbed on the versions that don't support them. It's in conflict with the design of OSContainer to only contain metrics which are commonly supported (should do the same on all versions). On the JDK side those version specific metrics live in version specific classes. Not sure if this design applies well to hotspot code.
- This is untested code. If anything, please add tests for this to `TestMisc.java` which uses PrintContainerInfo to exercise this code.
- What's the use-case for those diagnostics? Usually on kubernetes swap is disabled so the swap info there is quite useless in that case. Why is the kernel memory info needed? The bug doesn't say *why* it's important to have.

**Comment by MBaesken** [comment]
> A couple of comments:
> 
> * This introduces a couple of version specific info which is stubbed on the versions that don't support them. It's in conflict with the design of OSContainer to only contain metrics which are commonly supported (should do the same on all versions). On the JDK side those version specific metrics live in version specific classes. Not sure if this design applies well to hotspot code.
> * This is untested code. If anything, please add tests for this to `TestMisc.java` which uses PrintContainerInfo to exercise this code.
> * What's the use-case for those diagnostics? Usually on kubernetes swap is disabled so the swap info there is quite useless in that case. Why is the kernel memory info needed? The bug doesn't say _why_ it's important to have.

Hi Severin, yes sure testing needs to be done, thanks for the advice regarding TestMisc.java.
Regarding use-case I think Thomas Stuefe can give the best feedback.

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp [hunk 1]
```diff
@@ -252,6 +252,14 @@ class CgroupSubsystem: public CHeapObj<mtInternal> {
     virtual jlong memory_and_swap_limit_in_bytes() = 0;
     virtual jlong memory_soft_limit_in_bytes() = 0;
     virtual jlong memory_max_usage_in_bytes() = 0;
+
+    virtual jlong kernel_memory_usage_in_bytes() = 0;
+    virtual jlong kernel_memory_limit_in_bytes() = 0;
+    virtual jlong kernel_memory_max_usage_in_bytes() = 0;
+
+    virtual jlong memory_swap_current_in_bytes() = 0;
+    virtual jlong memory_swap_max_limit_in_bytes() = 0;
```
Comment (jerboaa) [inline src/hotspot/os/linux/cgroupSubsystem_linux.hpp:261]: I'm really not fond of this increased API surface for properties which are only supported on one version, but not the other. The idea would be to use those for printing in `os_linux`, right? Could we not move the implementation detail to the version specific classes? I.e. define `virtual void print_version_specific_info(outputStream* st) = 0` here and in `osContainer_linux.hpp` then only call `OSContainer::print_version_specific_info(st)` in `os_linux` and be done with it? This would avoid the empty stub implementations in the version specific classes...

File: src/hotspot/os/linux/cgroupV2Subsystem_linux.cpp [hunk 2]
```diff
@@ -212,6 +234,20 @@ char* CgroupV2Subsystem::mem_limit_val() {
...
+
+jlong CgroupV2Subsystem::kernel_memory_limit_in_bytes() {
+  return OSCONTAINER_ERROR;
+}
+
+jlong CgroupV2Subsystem::kernel_memory_max_usage_in_bytes() {
+  return OSCONTAINER_ERROR;
+}
+
+
```
Comment (jerboaa) [inline src/hotspot/os/linux/cgroupV2Subsystem_linux.cpp:250]: These should really not be needed if we make the `print_version_specific_info(...)` function implementation defined.

File: src/hotspot/os/linux/os_linux.cpp [hunk 3]
```diff
@@ -2266,6 +2266,19 @@ bool os::Linux::print_container_info(outputStream* st) {
   print_container_helper(st, OSContainer::memory_usage_in_bytes(), "memory_usage_in_bytes");
   print_container_helper(st, OSContainer::memory_max_usage_in_bytes(), "memory_max_usage_in_bytes");
 
+  // the kmem (kernel memory) values are only available in cgroupv1
+  if (p_ct != NULL && strcmp(p_ct, "cgroupv1") == 0) {
+    print_container_helper(st, OSContainer::kernel_memory_usage_in_bytes(), "kernel_memory_usage_in_bytes");
```
Comment (jerboaa) [inline src/hotspot/os/linux/os_linux.cpp:2271]: We wouldn't need the extra version checking gymnastics if we'd call an `OSContainer::print_version_specific_info(st);` here instead.

File: test/hotspot/jtreg/containers/docker/TestMisc.java [hunk 4]
```diff
@@ -123,6 +123,19 @@ private static void checkContainerInfo(OutputAnalyzer out) throws Exception {
         for (String s : expectedToContain) {
             out.shouldContain(s);
         }
+        String str = out.getOutput();
+        if (str.contains("cgroupv1")) {
```
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/TestMisc.java:127]: An alternative would be to pass a `Metrics` instance (`Metrics.systemMetrics()`) to `checkContainerInfo` and use:
`"cgroupv1".equals(m.getProvider())` here instead.

**Review by jerboaa — CHANGES_REQUESTED** [review]
Please consider pushing implementation details of the version specific printing to the implementation classes.

**Comment by iklam** [comment]
I think the issue title should be changed to "Improve container information", or maybe more specifically to  "Improve container information in hs_err and logging".

**Comment by MBaesken** [comment]
> I think the issue title should be changed to "Improve container information", or maybe more specifically to "Improve container information in hs_err and logging".

Thanks for the comment; I adjusted the title.

File: src/hotspot/os/linux/os_linux.cpp (hunk 3 above)
Comment (MBaesken) [reply src/hotspot/os/linux/os_linux.cpp:2271]: sounds like a good idea, but with the kmem stuff deprecated (Kernel commit adding the deprecation:
https://github.com/torvalds/linux/commit/0158115f702b )  we should maybe wait first for a reply from Thomas why he is still interested in it.

Comment (jerboaa) [reply src/hotspot/os/linux/os_linux.cpp:2271]: OK.

**Comment by tstuefe** [comment]
Hi Severin,

> * What's the use-case for those diagnostics? Usually on kubernetes swap is disabled so the swap info there is quite useless in that case. Why is the kernel memory info needed? The bug doesn't say _why_ it's important to have.

Memory+swap (v1) or only swap (v2) limits are interesting to us since they may limit memory usage of a container the same way memory limit does. We are often confronted with unexplainable OOM kills or container kills; it makes sense to know the whole cgroup config, not only part of that.

File: src/hotspot/os/linux/cgroupSubsystem_linux.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/linux/cgroupSubsystem_linux.hpp:261]: > I'm really not fond of this increased API surface for properties which are only supported on one version, but not the other. The idea would be to use those for printing in `os_linux`, right? Could we not move the implementation detail to the version specific classes? I.e. define `virtual void print_version_specific_info(outputStream* st) = 0` here and in `osContainer_linux.hpp` then only call `OSContainer::print_version_specific_info(st)` in `os_linux` and be done with it? This would avoid the empty stub implementations in the version specific classes...

We can do something much simpler. We can just expose the memory controller path and let `os::print_container_info()` read the files directly. Like we do e.g. with /proc/meminfo.

The really useful thing OsContainer does is figuring out the controller path. Wrt to hs-err printing and VM.info, I am not so interested in its v1-v2-agnostic abstraction. I just want to know the values of the files in that directory, verbatim.

If `os::print_container_info` were to access those files directly, we would not need to discuss which values are deemed useful or not. We just can print whatever we want without broadening the API surface. And I don't have to do the mental gymnastics to backtrace the v1-v2-abstraction numbers to their real platform meaning.

In short, I think abstractions are useful to programmatically base decisions on hotspot and java code. But for error analysis, I prefer the raw values with their exact names.

**Comment by tstuefe** [comment]
As I wrote in a comment, if it is a problem expanding the OsContainer API and getting a good and useful v1-v2-abstraction, then let's not. 

All I really wanted was 1) have c-groups info in hs-err and Vm.Info and 2) have a clarification wrt different meaning of `memory_and_swap_limit_in_bytes()` for v1 vs v2. I can live without (2), and for 1), I can live with a solution where OsContainer just exposes the memory controller path and in os_linux.cpp we print out the various files we are interested in directly. I can also live with Matthias' patch, which I think is clean and looks good.

Cheers, Thomas

File: src/hotspot/os/linux/os_linux.cpp (hunk 3 above)
Comment (MBaesken) [reply src/hotspot/os/linux/os_linux.cpp:2283]: Hi Severin, I think the OSContainer::print_version_specific_info(st) sounds like a good idea. It would mean that we keep the API smaller and do not overload it with stuff that is in fact only available in v1  (or only available in v2).

**Comment by MBaesken** [comment]
I introduced a  method  OSContainer::print_version_specific_info, this avoids putting (in some version) unavailable metrics (like the kernel-memory functions) into all cgroups (v1 and v2).

**Review by jerboaa — APPROVED** [review]
OK

File: src/hotspot/os/linux/osContainer_linux.hpp [hunk 5]
```diff
@@ -43,6 +43,9 @@ class OSContainer: AllStatic {
 
  public:
   static void init();
+  static void print_version_specific_info(outputStream* st);
+  static void print_container_helper(outputStream* st, jlong j, const char* metrics);
+
```
Comment (tstuefe) [inline src/hotspot/os/linux/osContainer_linux.hpp:48]: hpp file needs forward declaration of outputStream now, or include ostream.hpp, otherwise you break files that include this header but not ostream.

Comment (MBaesken) [reply src/hotspot/os/linux/osContainer_linux.hpp:49]: Hi Thomas, I added the inclusion of ostream.hpp .
Thanks, Matthias

**Review by tstuefe — APPROVED** [review]
Thank you Matthias. This looks good to me now.
